File: parser/src/anomaly.ts
/** ALP anomaly detection (v7.2.0 — Adaptive Policy & Continuous Governance). */

import { StoredEvent } from './state-store';

export interface AnomalyResult {
  event: StoredEvent;
  anomalies: string[];
  z_threshold: number;
  detected_at?: string;
}

export class AnomalyDetector {
  private baselines: Record<string, any>;
  private zThreshold: number;

  constructor(events: StoredEvent[], zThreshold = 3.0) {
    this.zThreshold = zThreshold;
    this.baselines = this.computeBaselines(events);
  }

  private computeBaselines(events: StoredEvent[]): Record<string, any> {
    const failureRates: number[] = [];
    const handoffRates: number[] = [];
    const agentClaims: Record<string, number> = {};
    const agentFailures: Record<string, number> = {};

    for (const e of events) {
      if (e.type === 'task_status' && e.status === '[!]') failureRates.push(1);
      else if (e.type === 'task_status') failureRates.push(0);
      if (e.type === 'human_handoff') handoffRates.push(1);
      if (e.agent) {
        agentClaims[e.agent] = (agentClaims[e.agent] || 0) + 1;
        if (e.type === 'workflow_fail' || e.status === '[!]') {
          agentFailures[e.agent] = (agentFailures[e.agent] || 0) + 1;
        }
      }
    }

    return {
      failure_rate_mean: mean(failureRates),
      failure_rate_stddev: stddev(failureRates),
      handoff_rate_mean: mean(handoffRates),
      handoff_rate_stddev: stddev(handoffRates),
      agent_claims: agentClaims,
      agent_failures: agentFailures,
    };
  }

  detect(event: StoredEvent): AnomalyResult | undefined {
    const anomalies: string[] = [];
    const { failure_rate_mean, failure_rate_stddev, handoff_rate_mean, handoff_rate_stddev, agent_claims, agent_failures } = this.baselines;

    if (event.type === 'task_status' && event.status === '[!]') {
      if (failure_rate_stddev > 0) {
        const z = Math.abs(1 - failure_rate_mean) / failure_rate_stddev;
        if (z > this.zThreshold) anomalies.push('failure_spike');
      } else if (failure_rate_mean === 0) {
        anomalies.push('failure_spike');
      }
    }

    if (event.type === 'human_handoff') {
      if (handoff_rate_stddev > 0) {
        const z = Math.abs(1 - handoff_rate_mean) / handoff_rate_stddev;
        if (z > this.zThreshold) anomalies.push('handoff_spike');
      }
    }

    if (event.agent) {
      const claims = agent_claims[event.agent] || 0;
      const failures = agent_failures[event.agent] || 0;
      if (claims > 0 && failure_rate_stddev > 0) {
        const ratio = failures / claims;
        const z = Math.abs(ratio - failure_rate_mean) / failure_rate_stddev;
        if (z > this.zThreshold) anomalies.push('agent_failure_rate');
      }
    }

    if (anomalies.length === 0) return undefined;
    return { event, anomalies, z_threshold: this.zThreshold, detected_at: event.timestamp };
  }

  detectBatch(events: StoredEvent[]): AnomalyResult[] {
    return events.map((e) => this.detect(e)).filter((r): r is AnomalyResult => r !== undefined);
  }

  updateThreshold(zThreshold: number): void {
    this.zThreshold = zThreshold;
  }
}

function mean(values: number[]): number {
  if (!values.length) return 0;
  return values.reduce((s, n) => s + n, 0) / values.length;
}

function stddev(values: number[]): number {
  if (values.length < 2) return 0;
  const m = mean(values);
  const variance = values.reduce((s, n) => s + (n - m) ** 2, 0) / (values.length - 1);
  return Math.sqrt(variance);
}

File: sdk/python/alp_sdk/anomaly.py
"""ALP anomaly detection (v7.2.0 — Adaptive Policy & Continuous Governance).

Detects statistical anomalies in agent and task behavior using
baseline profiles computed from historical event data.
"""

import math
from typing import Any, Dict, List, Optional


class AnomalyDetector:
    """Statistical anomaly detector for runtime events.

    Computes baselines from a training window of events and flags
    new observations that deviate beyond a configurable z-score
    threshold.
    """

    def __init__(self, events: List[Dict[str, Any]], z_threshold: float = 3.0):
        self.events = events
        self.z_threshold = z_threshold
        self.baselines = self._compute_baselines()

    def _compute_baselines(self) -> Dict[str, Any]:
        task_cycles: List[float] = []
        task_failures: List[int] = []
        task_handoffs: List[int] = []
        agent_claims: Dict[str, int] = {}
        agent_failures: Dict[str, int] = {}

        for e in self.events:
            etype = e.get("type", "")
            tid = e.get("task_id")
            agent = e.get("agent")
            if etype == "task_status" and tid:
                status = e.get("status")
                if status == "[!]":
                    task_failures.append(1)
                else:
                    task_failures.append(0)
            if etype == "human_handoff" and tid:
                task_handoffs.append(1)
            if agent:
                agent_claims[agent] = agent_claims.get(agent, 0) + 1
                if etype == "workflow_fail" or e.get("status") == "[!]":
                    agent_failures[agent] = agent_failures.get(agent, 0) + 1

        return {
            "task_cycle_mean": None,
            "task_cycle_stddev": None,
            "failure_rate_mean": _mean(task_failures) if task_failures else 0.0,
            "failure_rate_stddev": _stddev(task_failures) if task_failures else 0.0,
            "handoff_rate_mean": _mean(task_handoffs) if task_handoffs else 0.0,
            "handoff_rate_stddev": _stddev(task_handoffs) if task_handoffs else 0.0,
            "agent_claims": agent_claims,
            "agent_failures": agent_failures,
        }

    def detect(self, event: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Flag an event as anomalous if it exceeds the z-score threshold."""
        etype = event.get("type", "")
        agent = event.get("agent")
        anomalies: List[str] = []

        if etype == "task_status" and event.get("status") == "[!]":
            rate = self.baselines["failure_rate_mean"]
            std = self.baselines["failure_rate_stddev"]
            if std > 0:
                z = abs(1.0 - rate) / std
                if z > self.z_threshold:
                    anomalies.append("failure_spike")
            elif rate == 0.0:
                anomalies.append("failure_spike")

        if etype == "human_handoff":
            rate = self.baselines["handoff_rate_mean"]
            std = self.baselines["handoff_rate_stddev"]
            if std > 0 and rate > 0:
                z = (1.0 - rate) / std
                if z > self.z_threshold:
                    anomalies.append("handoff_spike")

        if agent:
            claims = self.baselines["agent_claims"].get(agent, 0)
            failures = self.baselines["agent_failures"].get(agent, 0)
            if claims > 0:
                failure_ratio = failures / claims
                baseline_failure = self.baselines["failure_rate_mean"]
                std = self.baselines["failure_rate_stddev"]
                if std > 0:
                    z = abs(failure_ratio - baseline_failure) / std
                    if z > self.z_threshold:
                        anomalies.append("agent_failure_rate")

        if anomalies:
            return {
                "event": event,
                "anomalies": anomalies,
                "z_threshold": self.z_threshold,
                "detected_at": event.get("timestamp"),
            }
        return None

    def detect_batch(self, events: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        results = [self.detect(e) for e in events]
        return [r for r in results if r is not None]

    def update_threshold(self, z_threshold: float) -> None:
        self.z_threshold = z_threshold


def _mean(values: List[float]) -> float:
    if not values:
        return 0.0
    return sum(values) / len(values)


def _stddev(values: List[float]) -> float:
    if len(values) < 2:
        return 0.0
    m = _mean(values)
    variance = sum((x - m) ** 2 for x in values) / (len(values) - 1)
    return math.sqrt(variance)
